recipes-openxt/qemu-dm/files/0027-nic-link-state-propagation.patch

```diff
@@ -0,0 +1,246 @@
++        fprintf(stderr, "Failed to find a network back-end for `%s'.\n", nc->name);
++        return;
++    }
++
++    dompath = xs_get_domain_path(xenstore, xen_domid);
++    if (!dompath) {
++        fprintf(stderr, "Could not retrieve domain path.\n");
++        return;
++    }
++    snprintf(base, sizeof (base), "%s/device/%s/%s", dompath, type, id);
++    free(dompath);
++
++    if (xenstore_read_int(base, "disconnect", &val)) {
++        fprintf(stderr, "Failed to read xenstore path (%s/%s).\n", base, "disconnect");
++        return;
++    }
++
++    /* The value is actually a boolean. */
++    if (nc->link_down != !!val) {
++        fprintf(stderr, "%s (%s%s): link status is now %s.\n", nc->name, type, id, !!val ? "down" : "up");
++        /* Notify the emulation through QMP.
++         * Note that qmp_set_link boolean is "link-up?", not nc->link_down "link-down?". */
++        qmp_set_link(nc->name, !val, NULL);
++    }
++}
++
++/* OpenXT: Register a Net Client in Xenstore. */
++int xenstore_register_nic(NetClientState *nc)
++{
++    char *dompath;
++    char type[XENSTORE_NET_TYPE_LEN];
++    char id[XENSTORE_NET_ID_LEN];
++    char base[XEN_BUFSIZE];
++
++    if (xenstore_net_client_state_fetch_backend_info(nc, type, id)) {
++        fprintf(stderr, "Failed to find a network back-end for `%s'.\n", nc->name);
++        return -1;
++    }
++
++    dompath = xs_get_domain_path(xenstore, xen_domid);
++    if (!dompath) {
++        fprintf(stderr, "Could not retrieve domain path.\n");
++        return -1;
++    }
++    snprintf(base, sizeof (base), "%s/device/%s/%s", dompath, type, id);
++    free(dompath);
++
++    if (xenstore_add_watch(base, "disconnect", xenstore_update_nic, nc)) {
++        fprintf(stderr, "Could not install xenstore watch on path `%s/disconnect'.\n", base);
++        return -1;
++    }
++
++    xenstore_update_nic(nc);
++
++    return 0;
++}
++
++/* OpenXT: Unregister a Net Client in Xenstore.
++ * Called when a device is removed and no longer used. */
++int xenstore_unregister_nic(NetClientState *nc)
++{
++    char *dompath;
++    char type[XENSTORE_NET_TYPE_LEN];
++    char id[XENSTORE_NET_ID_LEN];
++    char base[XEN_BUFSIZE];
++
++    if (xenstore_net_client_state_fetch_backend_info(nc, type, id)) {
++        fprintf(stderr, "Failed to find a network back-end for `%s'.\n", nc->name);
++        return -1;
++    }
++
++    dompath = xs_get_domain_path(xenstore, xen_domid);
++    if (!dompath) {
++        fprintf(stderr, "Could not retrieve domain path.\n");
++        return -1;
++    }
++    snprintf(base, sizeof (base), "%s/device/%s/%s", dompath, type, id);
++    free(dompath);
++
++    if (xenstore_remove_watch(base, "disconnect", xenstore_update_nic, id)) {
++        fprintf(stderr, "Could not remove xenstore watch on path `%s/disconnect'.\n", base);
++        return -1;
++    }
++    /* TODO: xenstore_update_nic() and force disconnect if not already? */
```

Comment:
No I forgot to remove it.
It is not necessary as this is called from net/net.c qemu_del_nic() (link-state doesn't matter at that point). Even if we wanted to "disconnect" using that helper it would only be from the XenStore.